RavenDB-21344 Aligning how both Corax & Lucene handle full text searc…

ayende opened the pull request:
…h queries using wildcards

Better DebugView messages in the query plan

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21344 

### Additional description

We previously did nasty things to analyzed terms with wild cards
This worked, to an extent, as long as you didn't use a full text analyzer.

Now we apply the following logic for wild card / contains / prefix and ends with queries inside `search()`:

* If you are using `exact` analyzer, we'll use `exact`
* If you are using _any other_ analyzer, we'll use the `lower case` analyzer, and that is _it_. 

See also: https://ayende.com/blog/191841-B/understanding-query-processing-and-wildcards-in-ravendb

ayende committed:
RavenDB-21344 Aligning how both Corax & Lucene handle full text search queries using wildcards
Better DebugView messages in the query plan

ppekrol commented on src/Corax/Analyzers/Analyzer.cs:
```diff
@@ -75,6 +76,8 @@ public static ArrayPool<Token> TokensPool
             float tokenBufferMultiplier = 1;
             foreach( var transformer in transformers)
             {
+                IsExactAnalyzer = transformer is ExactTransformer;
```
this sets that according to last transformer. Is this correct?

ayende replied:
We only use 1 right now, but you are correct WRT future. Fixed.

ayende committed:
RavenDB-21344 Ensure we future proof the IsExactAnalyzer and detect if there are any non exact analyzers

redknightlois commented on src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Common.cs:
```diff
@@ -37,6 +38,7 @@ private MultiTermMatch MultiTermMatchBuilder<TTermProvider>(in FieldMetadata fie
             seekKey = _fieldsTree.Llt.AcquireCompactKey();
             seekKey.Set(seekTerm.AsReadOnlySpan());
+            seekKey.ChangeDictionary(terms.DictionaryId);
         }
```
Why do we need to change the dictionary or better phrased, if that is needed, why it didn't fail before?

ayende replied:
This is so the `Inspect` call will tell us what is the actual key that is being used in terms.
That shouldn't cause any affect unless you are using query timings.

redknightlois commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -408,7 +408,7 @@ private static IQueryMatch ToCoraxQuery(Parameters builderParameters, QueryExpre
                     builderParameters.BuildSteps?.Add(
-                        $"OR operator: left - {left.GetType().FullName} ({left}) assembly: {left.GetType().Assembly.FullName} assemby location: {left.GetType().Assembly.Location} , right - {right.GetType().FullName} ({right}) assemlby: {right.GetType().Assembly.FullName} assemby location: {right.GetType().Assembly.Location}");
+                        $"OR operator: left - {left.GetType().FullName} ({left}) assembly: {left.GetType().Assembly.FullName} assembly location: {left.GetType().Assembly.Location} , right - {right.GetType().FullName} ({right}) assemlbly: {right.GetType().Assembly.FullName} assembly location: {right.GetType().Assembly.Location}");
 
```
Typo on assembly.

redknightlois commented:
LGTM

ayende committed:
RavenDB-21344
- When user is using a custom analyzer in Lucene, we'll use that even in wildcard queries, for backward compact. For Corax, we'll always use exact / lower mode.
- Fixing full text search *without* wildcards can now handle multiple terms from the analyzer in Corax
- Fixing typo

arekpalinski commented:
test this please

arekpalinski commented:
Note that we change Lucene behavior as well. @ppekrol @maciejaszyk  please review.

I'm merging it for testing.
